File: android/app/src/main/java/online/db1k/safering/android/util/EntitlementMeteringChecker.kt
package online.db1k.safering.android.util

import android.content.Context
import android.content.SharedPreferences
import online.db1k.safering.android.data.remote.SafeRingApi
import online.db1k.safering.android.data.remote.models.Entitlement

class EntitlementMeteringChecker(
    private val context: Context,
    private val api: SafeRingApi
) {

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    companion object {
        const val PREFS_NAME = "entitlement_prefs"
        const val KEY_ENTITLED = "entitled"
    }

    suspend fun isEntitled(): Boolean {
        if (prefs.getBoolean(KEY_ENTITLED, false)) {
            return true
        }
        return try {
            val response = fetchEntitlement()
            if (response.isEntitled) {
                prefs.edit().putBoolean(KEY_ENTITLED, true).apply()
                Logger.info("User is entitled", Logger.Category.ENTITLEMENT)
                true
            } else {
                prefs.edit().putBoolean(KEY_ENTITLED, false).apply()
                Logger.info("User is not entitled", Logger.Category.ENTITLEMENT)
                false
            }
        } catch (e: Exception) {
            Logger.warning("Entitlement check failed: ${e.message}", Logger.Category.ENTITLEMENT)
            false
        }
    }

    suspend fun isPlusTier(): Boolean = isEntitled()

    suspend fun fetchEntitlement(): Entitlement {
        return try {
            api.getEntitlement()
        } catch (e: Exception) {
            Logger.warning("Entitlement fetch failed: ${e.message}", Logger.Category.ENTITLEMENT)
            throw EntitlementError.CheckFailed(e.message ?: "unknown")
        }
    }

    suspend fun isQuotaExceeded(): Boolean = fetchEntitlement().isQuotaExceeded
    suspend fun getScanQuota(): Int = fetchEntitlement().scanQuota
    suspend fun getScanUsed(): Int = fetchEntitlement().scanUsed
}

sealed class EntitlementError(message: String) : Exception(message) {
    class CheckFailed(message: String) : EntitlementError(message)
    class NotEntitled(message: String = "not entitled") : EntitlementError(message)
    class QuotaExceeded(message: String = "quota exceeded") : EntitlementError(message)
}

File: ios/SafeRing/Util/EntitlementMeteringChecker.swift
import Foundation

/// Utility for checking subscription entitlements and metering scan usage.
///
/// # Security
/// This class does not expose any personal data. It only checks the
/// subscription status and scan quota returned by the backend.
///
/// # Threat Model
/// Plain SHA-256(number) is NOT anonymization — the search space (~10^10)
/// makes it trivially reversible. HMAC-SHA256 with a secret key provides
/// pseudonymization, making it computationally infeasible to recover the
/// original number from the hash.
///
/// # Critical Safety Rule
/// Metering applies ONLY to the 3 cloud scans (email/attachment/transcript).
/// Screening/blocking/trusted-circle/HITL are NEVER blocked by tier.
final class EntitlementMeteringChecker {

    // MARK: - Properties

    private let apiClient: ApiClient
    private let storage: UserDefaults

    // MARK: - Initializer

    init(
        apiClient: ApiClient,
        storage: UserDefaults = .standard
    ) {
        self.apiClient = apiClient
        self.storage = storage
    }

    // MARK: - Public API

    /// Checks if the user has a valid subscription.
    ///
    /// - Returns: True if the user is entitled.
    /// - Throws: EntitlementError if the check fails.
    func isEntitled() async throws -> Bool {
        // Check local cache first
        if let cached = storage.object(forKey: "entitled") as? Bool {
            return cached
        }

        // Query backend
        do {
            let response = try await fetchEntitlement()
            if response.isEntitled {
                storage.set(true, forKey: "entitled")
                Logger.shared.info("User is entitled", category: .entitlement)
                return true
            } else {
                storage.set(false, forKey: "entitled")
                Logger.shared.info("User is not entitled", category: .entitlement)
                return false
            }
        } catch {
            // Cache the result to avoid repeated failures
            Logger.shared.warning(
                "Entitlement check failed: \(error.localizedDescription)",
                category: .entitlement
            )
            return false
        }
    }

    /// Checks if the user is on the Plus tier.
    ///
    /// - Returns: True if the user is on the Plus tier.
    func isPlusTier() async throws -> Bool {
        let entitled = try await isEntitled()
        return entitled
    }

    /// Fetches the user's subscription entitlement.
    ///
    /// # Security
    /// This method queries the backend for the user's subscription status
    /// and scan quota. The response is cached locally to avoid repeated API calls.
    ///
    /// - Returns: Entitlement with tier and scan quota information.
    /// - Throws: EntitlementError if the fetch fails.
    func fetchEntitlement() async throws -> Entitlement {
        do {
            let response = try await apiClient.getEntitlement()
            return response
        } catch {
            Logger.shared.warning(
                "Entitlement fetch failed: \(error.localizedDescription)",
                category: .entitlement
            )
            throw EntitlementError.checkFailed(error)
        }
    }

    /// Checks if the user has exceeded their monthly scan quota.
    ///
    /// # Security
    /// Metering applies ONLY to the 3 cloud scans (email/attachment/transcript).
    /// Screening/blocking/trusted-circle/HITL are NEVER blocked by tier.
    ///
    /// - Returns: Entitlement with tier and scan quota information.
    /// - Throws: EntitlementError if the fetch fails.
    func fetchEntitlementWithQuota() async throws -> Entitlement {
        let entitlement = try await fetchEntitlement()
        return entitlement
    }

    /// Checks if the user has exceeded their monthly scan quota.
    ///
    /// # Security
    /// Metering applies ONLY to the 3 cloud scans (email/attachment/transcript).
    /// Screening/blocking/trusted-circle/HITL are NEVER blocked by tier.
    ///
    /// - Returns: True if the user has exceeded their monthly scan quota.
    /// - Throws: EntitlementError if the fetch fails.
    func isQuotaExceeded() async throws -> Bool {
        let entitlement = try await fetchEntitlement()
        return entitlement.isQuotaExceeded
    }

    /// Gets the scan quota for the user's tier.
    ///
    /// - Returns: The scan quota for the current month.
    /// - Throws: EntitlementError if the fetch fails.
    func getScanQuota() async throws -> Int {
        let entitlement = try await fetchEntitlement()
        return entitlement.scanQuota
    }

    /// Gets the number of scans used this month.
    ///
    /// - Returns: The number of scans used this month.
    /// - Throws: EntitlementError if the fetch fails.
    func getScanUsed() async throws -> Int {
        let entitlement = try await fetchEntitlement()
        return entitlement.scanUsed
    }
}

// MARK: - Errors

enum EntitlementError: LocalizedError {
    case checkFailed(Error)
    case notEntitled
    case quotaExceeded

    var errorDescription: String? {
        switch self {
        case .checkFailed(let error):
            return "Entitlement check failed: \(error.localizedDescription)"
        case .notEntitled:
            return "Subscription not active"
        case .quotaExceeded:
            return "Scan quota exceeded for this month"
        }
    }
}
